handler.go

```diff
@@ -0,0 +1,614 @@
+		// corresponding to the regular and the wildcard DNS names.
+		var wg sync.WaitGroup
+
+		// If regular DNS name info exists then update the corresponding DNSNameResolver CR for
+		// the DNS lookup failure.
+		if regularDNSExists {
+			wg.Add(1)
+			go func() {
+				defer wg.Done()
+				resolver.updateResolvedNamesFailure(ctx, regularDnsInfo, qname, status)
+			}()
+		}
+
+		// If wildcard DNS name info exists then update the corresponding DNSNameResolver CR for
+		// the DNS lookup failure.
+		if wildcardDNSExists {
+			wg.Add(1)
+			go func() {
+				defer wg.Done()
+				resolver.updateResolvedNamesFailure(ctx, wildcardDnsInfo, qname, status)
+			}()
+		}
+
+		// Wait for the goroutines to complete.
+		wg.Wait()
+
+		// Return the response received from the plugin chain.
+		return status, err
+	}
+
+	// Get the IP addresses and the corresponding TTLs in a map. Only A and AAAA type DNS records
+	// are considered.
+	ipTTLs := make(map[string]int32)
+	for _, answer := range rw.Msg.Answer {
+		switch state.QType() {
+		case dns.TypeA:
+			if rec, ok := answer.(*dns.A); ok {
+				ttl := int32(rec.Hdr.Ttl)
+				if ttl == 0 {
+					ttl = resolver.minimumTTL
+				}
+				ipTTLs[rec.A.String()] = ttl
+			}
+		case dns.TypeAAAA:
+			if rec, ok := answer.(*dns.AAAA); ok {
+				ttl := int32(rec.Hdr.Ttl)
+				if ttl == 0 {
+					ttl = resolver.minimumTTL
+				}
+				ipTTLs[rec.AAAA.String()] = ttl
+			}
+		default:
+			return status, err
+		}
+	}
+
+	// If no IP address is Return the response received from the plugin chain.
+	if len(ipTTLs) == 0 {
+		return status, err
+	}
+
+	// WaitGroup variable used to wait for the completion of update of DNSNameResolver CRs
+	// corresponding to the regular and the wildcard DNS names.
+	var wg sync.WaitGroup
+
+	// If regular DNS name info exists then update the corresponding DNSNameResolver CR for
+	// the successful DNS lookup.
+	if regularDNSExists {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			resolver.updateResolvedNamesSuccess(ctx, regularDnsInfo, qname, ipTTLs)
+
+		}()
+	}
+
+	// If wildcard DNS name info exists then update the corresponding DNSNameResolver CR for
+	// the successful DNS lookup.
+	if wildcardDNSExists {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			resolver.updateResolvedNamesSuccess(ctx, wildcardDnsInfo, qname, ipTTLs)
+		}()
+	}
+
+	// Wait for the goroutines to complete.
+	wg.Wait()
+
+	// Return the response received from the plugin chain.
+	return status, err
+}
+
+// Name implements the Handler interface.
+func (resolver *OCPDNSNameResolver) Name() string { return "ocp_dnsnameresolver" }
+
+// updateResolvedNamesSuccess updates the ResolvedNames field of the corresponding DNSNameResolver object when DNS lookup is successfully completed.
+func (resolver *OCPDNSNameResolver) updateResolvedNamesSuccess(ctx context.Context, namespaceDNS namespaceDNSInfo, dnsName string, ipTTLs map[string]int32) {
```

Comment:
This function is complex, I'm having hard times understanding the logic. The iteration over the resolved names with many booleans which are set and reset  in different `if`s and `else`s is the most difficult part. We need to do something to make it easier for the future maintenance. For instance, make each `if` case a dedicated function. The part after the `for` over the resolver names is complex too, may be deserving dedicated function(s) too.

Reply:
I will try to simplify the code by breaking them down into multiple functions.

Reply:
I have broken down the code into multiple functions. I hope the flow will be easier to understand with this simplification.